File: app/src/main/java/me/dev4vin/samplepromiseapp/ui/MainActivity.java
package me.dev4vin.samplepromiseapp.ui;

import android.os.Bundle;
import androidx.annotation.Nullable;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.Toolbar;
import android.view.View;
import android.widget.LinearLayout;

import me.dev4vin.samplepromiseapp.R;
import me.dev4vin.samplepromiseapp.common.DataRepository;
import me.dev4vin.samplepromiseapp.models.Todo;
import me.dev4vin.samplepromiseapp.ui.views.LoadableView;
import me.yoctopus.model.List;
import me.yoctopus.model.ResponseCallBack;
import me.yoctopus.view.AdapterDivider;
import me.yoctopus.view.PromiseAdapter;
import me.yoctopus.view.SearchableAdapter;
import me.yoctopus.view.loading.ProgressLayout;

public class MainActivity extends AppCompatActivity implements PromiseAdapter.Listener<Todo> {
  protected Toolbar toolbar;
  protected RecyclerView todosList;
  protected ProgressLayout loadingView;
  protected FloatingActionButton fab;
  private SearchableAdapter<Todo> searchableAdapter;
  private AdapterDivider divider;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    super.setContentView(R.layout.activity_main);
    initView();
    setSupportActionBar(toolbar);
  }

  @Override
  protected void onPostCreate(@Nullable Bundle savedInstanceState) {
    super.onPostCreate(savedInstanceState);
    fab.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
            .setAction("Action", null).show();
      }
    });

    divider = new AdapterDivider(this, LinearLayout.VERTICAL);
    searchableAdapter = new SearchableAdapter<>(this);
    todosList.setLayoutManager(new LinearLayoutManager(this));
    todosList.setItemAnimator(new DefaultItemAnimator());
    todosList.addItemDecoration(divider);
    loadingView.showLoading(new LoadableView("Loading todos, please wait..."));

    searchableAdapter.swipe(new PromiseAdapter.Swipe<Todo>() {
      @Override
      public void onSwipeRight(Todo todo, PromiseAdapter.Response response) {

      }

      @Override
      public void onSwipeLeft(Todo todo, PromiseAdapter.Response response) {

      }
    });
    todosList.setAdapter(searchableAdapter);
    DataRepository.instance().getTodos(0, 10, new ResponseCallBack<List<Todo>, Exception>()
    .response(new ResponseCallBack.Response<List<Todo>, Exception>() {
      @Override
      public void onResponse(List<Todo> todos) throws Exception {
        searchableAdapter.add(todos);
      }
    }).error(new ResponseCallBack.Error<Exception>() {
          @Override
          public void onError(final Exception e) {
            runOnUiThread(new Runnable() {
              @Override
              public void run() {
                loadingView.showEmpty(0, "Could not load todos", e.getMessage());
              }
            });
          }
        }));
  }

  @Override
  public void onClick(Todo todo, int id) {

  }

  private void initView() {
    toolbar = findViewById(R.id.toolbar);
    todosList = findViewById(R.id.todos_list);
    loadingView = findViewById(R.id.loading_view);
    fab = findViewById(R.id.fab);
  }
}

File: app/src/main/java/promise/app/ui/MainActivity.kt
package promise.app.ui

import android.os.Bundle
import android.widget.LinearLayout
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*
import promise.app.R
import promise.app.common.DataRepository
import promise.app.error.AppError
import promise.app.models.Todo
import promise.app.ui.base.BasePresenter
import promise.app.ui.base.BaseView
import promise.app.ui.base.PresenterActivity
import promise.cac.notif.Notification
import promise.model.List
import promise.model.ResponseCallBack
import promise.view.AdapterDivider
import promise.view.PromiseAdapter
import promise.view.SearchableAdapter

interface MainView : BaseView {
  fun onGetTodos(todos: List<Todo>)
  fun onError(error: AppError)
}

class MainPresenter(view: MainView) : BasePresenter<MainView>(view) {
  private val dataRepository: DataRepository by lazy { DataRepository.instance() }
  fun getTodos(skip: Int, take: Int) {
    dataRepository.getTodos(skip, take, ResponseCallBack<List<Todo>, Exception>()
        .response { view?.onGetTodos(it) }
        .error { view?.onError(AppError(it)) })
  }
}

class MainActivity : PresenterActivity<MainPresenter>(), PromiseAdapter.Listener<Todo>, MainView {

  override fun presenter(): MainPresenter = MainPresenter(this)

  override fun onGetTodos(todos: List<Todo>) {
    runOnUiThread {
      loading_view.showContent()
      notification.showDialog("Loaded todos", "they are: ${todos.size}", null, null, null)
    }
    searchableAdapter.add(todos)
  }

  override fun onError(error: AppError) {
    runOnUiThread {
      loading_view.showEmpty(0, "Could not load todos", error.message)
    }
  }

  private val searchableAdapter: SearchableAdapter<Todo> by lazy { SearchableAdapter<Todo>(this) }
  private  val divider: AdapterDivider by lazy { AdapterDivider(this, LinearLayout.VERTICAL) }
  private val notification: Notification by lazy { Notification(this) }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    super.setContentView(R.layout.activity_main)
    setSupportActionBar(toolbar)
  }

  override fun onPostCreate(savedInstanceState: Bundle?) {
    super.onPostCreate(savedInstanceState)
    todos_list.layoutManager = LinearLayoutManager(this)
    todos_list.itemAnimator = DefaultItemAnimator()
    todos_list.addItemDecoration(divider)
    loading_view.showLoading(/*LoadableView("Loading todos, please wait...")*/)

    searchableAdapter.swipe(object : PromiseAdapter.Swipe<Todo> {
      override fun onSwipeRight(todo: Todo, response: PromiseAdapter.Response) {

      }

      override fun onSwipeLeft(todo: Todo, response: PromiseAdapter.Response) {

      }
    })
    todos_list.adapter = searchableAdapter
    presenter.getTodos(0, 10)
    /*Handler().postDelayed( {
      notification.showDialog("Battery", " ${ApiHelper.batteryPercentage(this)}", null, null, null)
    }, 1000)*/
  }

  override fun onClick(todo: Todo, id: Int) {
      notification.showDialog("Todo clicked", todo.toString(),
          null, null, null)
  }

}
